Port to OpenBSD!

cfallin opened the pull request:
Verified to work with several simple Wasm modules on an OpenBSD 6.9
x86_64 machine, built with the Rust 1.51.0 included in the packages
collection (`pkg_add rust`) because `rustup` doesn't support
OpenBSD.

This requires a number of PRs to upstream crates to be merged and
released on crates.io first; as a temporary measure, this commit
includes a number of patches in the toplevel `Cargo.toml` to refer
to git branches. Once those patches are merged and released,
I can update this PR (cc @sunfishcode).

The only change to wasmtime code itself is in the trap-handler runtime.
I had to hardcode the offset to RIP in the signal frame, because `libc`
doesn't have a struct definition for OpenBSD. If we think it would be better
to get that upstreamed, I can do that instead. Or, we could go back to a
C helper and just include `<sys/signal.h>`, but I suspect we don't want to do that.

This also disables `wasi-nn` by default, because the `openvino-sys`
crate fails to build on OpenBSD. Ideally we would either fix
upstream or include some more build configuration to exclude the
feature on our end on OpenBSD; unfortunately it seems (according
to rust-lang/cargo#1197) that target-specific features are not
yet supported by Cargo, so this is not very easy to do. I'd be interested
in @abrown's thoughts on this!

Also note that we won't be able to build or test this platform on CI,
unfortunately, for three reasons: (i) `rustup` doesn't carry toolchain builds
for OpenBSD, I think because of the OS's "no ABI compatibility across releases"
policy; (ii) GitHub CI Actions doesn't have any runners
for this platform; and (iii) distributing binaries is difficult, for the above
ABI-compatibility reasons. Nevertheless, it would be nice to be able to say
that it has been known to work at one time, and also maybe at some point
in the distant future someone will see fit to include us in the OpenBSD
ports/packages tree.

Low priority for review (this was just a fun side-adventure). Thanks!

<!--

Please ensure that the following steps are all taken care of before submitting
the PR.

- [ ] This has been discussed in issue #..., or if not, please tell us why
  here.
- [ ] A short description of what this does, why it is needed; if the
  description becomes long, the matter should probably be discussed in an issue
  first.
- [ ] This PR contains test cases, if meaningful.
- [ ] A reviewer from the core maintainer team has been assigned for this PR.
  If you don't know who could review this, please indicate so. The list of
  suggested reviewers on the right can help you.

Please ensure all communication adheres to the [code of
conduct](https://github.com/bytecodealliance/wasmtime/blob/master/CODE_OF_CONDUCT.md).
-->

cfallin commented:
I'll also note that this depends on darfink/region-rs#8, which was recently added but has not yet been released yet; I'll ping over there to see if they can release a new version to crates.io.

abrown commented:
@cfallin, I'm interested in why the openvino-sys crate is failing: the way we build it here, it _should_ have no target-specific logic since it uses `libloading` to look for the right libraries. Can you send me the failed build output and I'll try to fix upstream?

cfallin commented:
> @cfallin, I'm interested in why the openvino-sys crate is failing: the way we build it here, it _should_ have no target-specific logic since it uses `libloading` to look for the right libraries. Can you send me the failed build output and I'll try to fix upstream?

It's an issue with some platform-specific constants that are defined with `#[cfg(linux)]` or similar rather than "all Unix-like platforms", I think:

```
error[E0425]: cannot find value `ENV_LIBRARY_PATH` in this scope
  --> /home/cfallin/.cargo/registry/src/github.com-1ecc6299db9ec823/openvino-finder-0.3.1/src/lib.rs:44:37
   |
44 |     if let Some(path) = env::var_os(ENV_LIBRARY_PATH) {
   |                                     ^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find value `DEFAULT_INSTALLATION_DIRECTORIES` in this scope
  --> /home/cfallin/.cargo/registry/src/github.com-1ecc6299db9ec823/openvino-finder-0.3.1/src/lib.rs:54:24
   |
54 |       for default_dir in DEFAULT_INSTALLATION_DIRECTORIES
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `KNOWN_INSTALLATION_SUBDIRECTORIES`
```

sunfishcode commented:
system-interface 0.6.6 and posish 0.6.3 are now available.

cfallin commented on crates/runtime/src/traphandlers/unix.rs:
```diff
@@ -158,6 +158,19 @@ unsafe fn get_pc(cx: *mut libc::c_void, _signum: libc::c_int) -> *const u8 {
+            // This is the kernel-entry register-save frame format and is
+            // *unlikely* to change across OpenBSD versions, but note that
+            // OpenBSD makes no guarantee of ABI compatibility across any
+            // release.
```
Sure thing -- created rust-lang/libc#2242. It might be a bit until this is easily usable without building rustc from source (the platform doesn't have rustup so it would likely be available with the packages in the next OpenBSD release in October) but then this PR isn't in any hurry either :-)

cfallin replied:
So it looks like there was an ongoing PR for this (rust-lang/libc#2189) that is somewhat complex, and stalled. For now I added a C helper and I'm happy to revisit this when Rust support exists, if that seems reasonable to you! (Also I misspoke above w.r.t. waiting for an OS package update -- `libc` is just a crate, of course, not part of the stdlib distro proper.)

sunfishcode replied:
The C wrapper looks fine to me. I suggest adding a link to rust-lang/libc#2189 in a comment so that people looking to refactor that code can easily check on its status.

abrown commented on Cargo.toml:
```diff
@@ -87,7 +87,7 @@ members = [
 ]
 
 [features]
-default = ["jitdump", "wasmtime/wat", "wasmtime/parallel-compilation", "wasi-nn"]
```
@cfallin, I fixed the `openvino` crates so the error you saw above should be resolved. I think this line should be unnecessary if you bump the `openvino` crate to v0.3.2.

cfallin replied:
Thanks!

cfallin commented:
@sunfishcode I've updated this now with some spare cycles and I think it might be good for another look, whenever you have the chance! It's still pending a region.rs release but I've pinged over there. I resolved the `libc` dependency for now with a C helper, which isn't ideal but is better than a hardcoded constant.

There's also some version weirdness where cap-std-* is pulling in two versions of `rustc_version` (see cargo-deny CI job); not sure what's going on there.

cfallin commented on crates/runtime/src/traphandlers/unix.rs:
```diff
@@ -127,6 +127,17 @@ unsafe extern "C" fn trap_handler(
     }
 }
 
+#[cfg(target_os = "openbsd")]
```
Done!

cfallin commented on .github/workflows/main.yml:
```diff
@@ -58,7 +58,7 @@ jobs:
     - uses: ./.github/actions/install-rust
       with:
-        toolchain: nightly-2021-04-11
+        toolchain: nightly-2021-06-27
```
Ah, yes, I had missed that, thanks!

cfallin commented:
Updated, thanks! This just depends now on (i) a region-rs release, and (ii) resolution of some multi-crate-version weirdness:

@sunfishcode, it seems that we have:
- system-interface 0.6.6 -> cap-fs-ext 0.13.7 -> unsafe_io 0.6.12 -> rustc_version 0.3.3
- system-interface 0.6.6 -> cap-fs-ext 0.13.7 -> unsafe_io 0.6.12 -> io_lifetimes 0.1.1 -> rustc_version 0.4.0
- system-interface 0.6.6 -> cap-fs-ext 0.13.10 -> rustc_version 0.3.3
- system-interface 0.6.6 -> rustc_version 0.4.0

in general it seems the various cap-std-related crates have inconsistent dependencies on `rustc_version`, which is making cargo-deny unhappy -- thoughts? Would it be possible to harmonize everything to 0.4.0?

cfallin commented:
@alexcrichton PTAL again (with low priority for this unofficial side-project!) re: fibers -- there's some additional flags-setting needed that I hadn't realized before too, as OpenBSD has a special mitigation that requires blessing stack memory in a certain way.

bjorn3 commented on crates/fiber/src/unix.rs:
```diff
@@ -74,22 +74,18 @@ unsafe fn mmap_new_stack(size: usize, page_size: usize) -> io::Result<(*mut u8,
     // stack; otherwise, we will get a SIGSEGV when it is used as such.
     //
-    // Note also that MAP_STACK must be specified with PROT_READ|PROT_WRITE, and
+    // Note that MAP_STACK must be specified with PROT_READ|PROT_WRITE, and
```
`MAP_STACK` can also be used on Linux. It is currently a no-op, but according to the man page it is accepted such that if a future architecture requires special handling for the stack, no program changes are necessary.

cfallin commented:
With latest updates, all tests pass in `wasmtime` crate. I had to disable the pooling allocator as well, for `MAP_STACK`-related reasons. Checking all tests would require me to set up a wasm32-wasi toolchain, which means building rustc from source (whee!), so that's the last of my side-project steam for now.

On a slightly less-unofficial note, at some point it would be neat to articulate a tier strategy for platform support; I imagine something like Rust's tier 1/2/3 (tested, compiled, and best-effort respectively?), where OpenBSD is definitely tier 3, but there might be some systems we could include at tier 2 (other BSDs? Linux/musl? macOS/aarch64?) if cross-compilation works.

bjorn3 commented on crates/runtime/src/instance/allocator/pooling.rs:
```diff
@@ -344,6 +349,15 @@ impl InstancePool {
+        instance_limits: &InstanceLimits,
+        tunables: &Tunables,
+    ) -> Result<Self> {
+        panic!("Pooling allocator not supported on OpenBSD due to limits in stack mmap'ing.");
```
What kind of limits?

cfallin replied:
MAP_STACK and issues with the way that the pooling allocator deallocs/reallocs to zero; see the commit message on the commit that added this line.

cfallin replied:
To give a little more detail: [this check](https://github.com/openbsd/src/blob/d0b698e17cb070a8f3e5081524f33169e71390a0/sys/uvm/uvm_mmap.c#L258-L259) in the OpenBSD kernel's mmap() disallows requesting a fixed address for a MAP_STACK mmap, so one can't use this to get fresh CoW pages in an existing stack area.

alexcrichton commented on tests/all/async_functions.rs:
```diff
@@ -447,6 +448,7 @@ fn async_with_pooling_stacks() {
 }
 
 #[test]
+#[cfg(not(target_os = "openbsd"))]
```
Could this (and the above test) use `skip_pooling_allocator_tests` instead?

alexcrichton commented on tests/all/limits.rs:
```diff
@@ -220,6 +220,7 @@ fn test_initial_table_limits_exceeded() -> Result<()> {
 }
 
 #[test]
+#[cfg(not(target_os = "openbsd"))]
```
Can this use `skip_pooling_allocator_tests`?

alexcrichton commented on tests/all/traps.rs:
```diff
@@ -539,6 +547,7 @@ fn parse_dwarf_info() -> Result<()> {
 
 #[test]
 #[cfg_attr(all(target_os = "macos", target_arch = "aarch64"), ignore)] // TODO #2808 system libunwind is broken on aarch64
+#[cfg_attr(target_os = "openbsd", ignore)] // backtraces are broken on OpenBSD
```
Could this perhaps be extracted to a helper function in this module that tests for arm64 macos and openbsd and skips the tests if that's true?

sunfishcode commented:
@cfallin https://github.com/bytecodealliance/wasmtime/pull/3049 updates almost everything to rustc_version 0.4.0, except wasmtime still depends on criterion which depends on cast which depends on rustc_version 0.3.3. I've now submitted https://github.com/japaric/cast.rs/pull/34 to update cast.

cfallin commented:
I just rebased onto latest `main` and updated to region-rs 3.0.0, which includes a fix for OpenBSD.

Unfortunately it looks like new build failures have cropped up in the meantime, this time in `rsix` (bytecodealliance/rsix#32). I also had to upgrade my test VM to OpenBSD-current to get Rust 1.54.0 because -stable (currently 6.9) has Rust 1.51.0, which fails to build some other things. (Rustup doesn't support OpenBSD unfortunately so one must use the system package or build from source.)

Given the moving target, the fact that we probably won't have a reasonable way to test this on CI, and the fact that we (or at least I) don't really have the time to support another platform beyond "one weekend morning once every few months" (this PR so far), I think I'll go ahead and close this PR. We should definitely reconsider this if/when OpenBSD graduates from tier3 and Rustup adds binaries, and if/when we have a way of adding native GitHub CI runners, at which point someone could run OpenBSD CI.

Until then I'll keep the `wasmtime` binary on my little router as a keepsake from that one golden moment when this PR compiled cleanly. Thanks for the patience, all!
